File: cpp/src/java/lang/StringBuilder.cpp
#include <cstring>

#include "java/lang/StringBuilder.h"

namespace java {

void
StringBuilder::ensureCapacity(int requiredLength) {
    if ( requiredLength + 1 <= capacity ) {
        return;
    }
    int newCapacity = capacity <= 0 ? 16 : capacity;
    while ( newCapacity < requiredLength + 1 ) {
        newCapacity *= 2;
    }
    char *newValue = new char[static_cast<std::size_t>(newCapacity)];
    if ( value == nullptr ) {
        newValue[0] = '\0';
    } else {
        std::strcpy(newValue, value);
        delete[] value;
    }
    value = newValue;
    capacity = newCapacity;
}

void
StringBuilder::assignFromCString(const char *text) {
    const char *source = text == nullptr ? "" : text;
    const int sourceLength = static_cast<int>(std::strlen(source));
    ensureCapacity(sourceLength);
    std::strcpy(value, source);
    lengthValue = sourceLength;
}

StringBuilder::StringBuilder():
    value(nullptr),
    lengthValue(0),
    capacity(0)
{
    ensureCapacity(0);
    value[0] = '\0';
}

StringBuilder::StringBuilder(const StringBuilder &other):
    value(nullptr),
    lengthValue(0),
    capacity(0)
{
    assignFromCString(other.value);
}

StringBuilder::StringBuilder(const String &text):
    value(nullptr),
    lengthValue(0),
    capacity(0)
{
    assignFromCString(text.toCString());
}

StringBuilder::~StringBuilder() {
    dispose();
}

void
StringBuilder::dispose() {
    if ( value != nullptr ) {
        delete[] value;
        value = nullptr;
    }
    lengthValue = 0;
    capacity = 0;
}

StringBuilder &
StringBuilder::operator=(const StringBuilder &other) {
    if ( this != &other ) {
        assignFromCString(other.value);
    }
    return *this;
}

int
StringBuilder::length() const {
    return lengthValue;
}

char
StringBuilder::charAt(int index) const {
    if ( index < 0 || index >= lengthValue ) {
        return '\0';
    }
    return value[index];
}

void
StringBuilder::clear() {
    lengthValue = 0;
    if ( value != nullptr ) {
        value[0] = '\0';
    }
}

StringBuilder &
StringBuilder::append(const String &text) {
    return append(text.toCString());
}

StringBuilder &
StringBuilder::append(const char *text) {
    if ( text == nullptr ) {
        return *this;
    }
    const int textLength = static_cast<int>(std::strlen(text));
    if ( textLength <= 0 ) {
        return *this;
    }
    ensureCapacity(lengthValue + textLength);
    std::strcat(value, text);
    lengthValue += textLength;
    return *this;
}

StringBuilder &
StringBuilder::append(const char *text, int textLength) {
    if ( text == nullptr || textLength <= 0 ) {
        return *this;
    }
    ensureCapacity(lengthValue + textLength);
    std::strncat(value, text, static_cast<std::size_t>(textLength));
    lengthValue += textLength;
    value[lengthValue] = '\0';
    return *this;
}

StringBuilder &
StringBuilder::append(char ch) {
    ensureCapacity(lengthValue + 1);
    value[lengthValue] = ch;
    lengthValue++;
    value[lengthValue] = '\0';
    return *this;
}

String
StringBuilder::toString() const {
    return String(value);
}

}

File: typescript/src/main/java/lang/StringBuilder.ts
import { String } from "./String";

export class StringBuilder {
  private value: string;

  public constructor(text?: StringBuilder | String) {
    if (text instanceof StringBuilder) {
      this.value = text.toString();
      return;
    }
    if (text instanceof String) {
      this.value = text.toCString();
      return;
    }
    this.value = "";
  }

  public dispose(): void {
    this.value = "";
  }

  public assign(other: StringBuilder): StringBuilder {
    this.value = other.toString();
    return this;
  }

  public length(): number {
    return this.value.length;
  }

  public charAt(index: number): string {
    if (index < 0 || index >= this.value.length) {
      return "\0";
    }
    return this.value.charAt(index);
  }

  public clear(): void {
    this.value = "";
  }

  public append(text: String | string): StringBuilder;
  public append(text: string, textLength: number): StringBuilder;
  public append(text: String | string, textLength?: number): StringBuilder {
    const raw = text instanceof String ? text.toCString() : text;
    if (!raw || raw.length === 0) {
      return this;
    }
    if (textLength === undefined) {
      this.value += raw;
      return this;
    }
    if (textLength > 0) {
      this.value += raw.slice(0, textLength);
    }
    return this;
  }

  public appendChar(ch: string): StringBuilder {
    if (!ch || ch.length === 0) {
      return this;
    }
    this.value += ch.charAt(0);
    return this;
  }

  public toStringObject(): String {
    return new String(this.value);
  }

  public toString(): string {
    return this.value;
  }
}
